File: src/main/scala/org/galaxio/gatling/amqp/request/ConsumeDslBuilder.scala
package org.galaxio.gatling.amqp.request

import io.gatling.core.action.builder.ActionBuilder
import io.gatling.core.session.Expression
import org.galaxio.gatling.amqp.AmqpCheck
import org.galaxio.gatling.amqp.action.ConsumeBuilder
import io.gatling.core.config.GatlingConfiguration

case class ConsumeDslBuilderQueue(requestName: Expression[String], configuration: GatlingConfiguration) {
  def queue(name: Expression[String]): ConsumeDslBuilder =
    ConsumeDslBuilder(ConsumeAttributes(requestName, name), configuration)
}

case class ConsumeDslBuilder(attributes: ConsumeAttributes, configuration: GatlingConfiguration) {
  def timeout(millis: Long): ConsumeDslBuilder = copy(attributes = attributes.copy(timeout = millis))

  def check(checks: AmqpCheck*): ConsumeDslBuilder =
    copy(attributes = attributes.copy(checks = attributes.checks ::: checks.toList))

  def silent: ConsumeDslBuilder = copy(attributes = attributes.copy(silent = true))

  def build(): ActionBuilder = ConsumeBuilder(attributes, configuration)
}

File: src/main/java/org/galaxio/gatling/amqp/javaapi/request/ConsumeDslBuilder.java
package org.galaxio.gatling.amqp.javaapi.request;

import io.gatling.javaapi.core.ActionBuilder;
import org.galaxio.gatling.amqp.javaapi.check.AmqpChecks;

import java.util.Arrays;
import java.util.List;

public class ConsumeDslBuilder implements ActionBuilder {
    private org.galaxio.gatling.amqp.request.ConsumeDslBuilder wrapped;

    public ConsumeDslBuilder(org.galaxio.gatling.amqp.request.ConsumeDslBuilder wrapped) {
        this.wrapped = wrapped;
    }

    public ConsumeDslBuilder timeout(Long millis) {
        this.wrapped = wrapped.timeout(millis);
        return this;
    }

    public ConsumeDslBuilder check(Object... checks) {
        return check(Arrays.asList(checks));
    }

    public ConsumeDslBuilder check(List<Object> checks) {
        this.wrapped = wrapped.check(AmqpChecks.toScalaChecks(checks));
        return this;
    }

    public ConsumeDslBuilder silent() {
        this.wrapped = wrapped.silent();
        return this;
    }

    public io.gatling.core.action.builder.ActionBuilder asScala() {
        return wrapped.build();
    }
}
